MarkBind/markbind
# Log website URL upon deployment

**What is the purpose of this pull request? (put "X" next to an item, remove the rest)**

• [ ] Documentation update
• [ ] Bug fix
• [ ] New feature
• [X] Enhancement to an existing feature
• [ ] Other, please explain:






**What is the rationale for this request?**
 This change helps the user easily access the website after deploying it using the MarkBind CLI.

**What changes did you make? (Give an overview)**
Github supports custom domains through the use of `CNAME`. However, to do so a file called `CNAME` must be placed in the root of the remote repository where we are deploying (the file is automatically created if one sets a manual deploy url through the github repo itself). Hence when `markbind deploy` is run, the following additional steps are done:
1. read the contents of the `CNAME` file in the `origin` remote with external library `simple-git`.
2. If not, there must be no custom url, and we proceed to manually construct the URL, first from the provided `repo` from `site.json`, otherwise using the remote itself.

**Provide some example code that this change will affect:**


```js
 static getDeploymentUrl(options) {
    const HTTPS_PART = 'https://';
    const SSH_PART = 'git@github.com:';
    const GITHUB_IO_PART = 'github.io';

    // https://<name|org name>.github.io/<repo name>/
    function constructGhPagesUrl(remoteUrl) {
      if (remoteUrl.includes(HTTPS_PART)) {
        // https://github.com/<name|org>/<repo>.git (HTTPS)
        const parts = remoteUrl.split('/');
        const repoName = parts[parts.length - 1].toLowerCase();
        const name = parts[parts.length - 2].toLowerCase();
        return `https://${name}.${GITHUB_IO_PART}/${repoName}`;
      } else if (remoteUrl.includes(SSH_PART)) {
        // git@github.com:<name|org>/<repo>.git (SSH)
        const parts = remoteUrl.split('/');
        const repoName = parts[parts.length - 1].toLowerCase();
        const name = (parts[0].split(':'))[0];
        return `https://${name}.${GITHUB_IO_PART}/${repoName}`;
      }
      throw new Error(`Unknown remote url ${remoteUrl}`);
    }

    const { remote, branch, repo } = options;
    const cnamePm = gitUtil.getRemoteBranchFile('blob', remote, branch, 'CNAME');
    const remoteUrlPm = gitUtil.getRemoteUrl(remote);
    const promises = [cnamePm, remoteUrlPm];

    return Promise.all(promises)
      .then((results) => {
        const cname = results[0];
        const remoteUrl = results[1];
        if (cname) {
          return cname;
        } else if (repo) {
          return constructGhPagesUrl(repo);
        }
        return constructGhPagesUrl(remoteUrl);
      })
      .catch((err) => {
        logger.error(err);
        return null;
      });
  }
```

**Testing instructions:**
1. Create a dummy website using `markbind init`.
2. Run `markbind deploy` and check the url the website is deployed at.

**Proposed commit message: (wrap lines at 72 characters)**
Log website URL upon deployment
